pkg/api/types.go

```diff
@@ -466,6 +470,30 @@ type GlusterfsVolumeSource struct {
 	ReadOnly bool `json:"readOnly,omitempty"`
 }
 
+// RBDVolumeSource represents a Rados Block Device Mount that lasts the lifetime of a pod
+type RBDVolumeSource struct {
+	// Required: EndpointsName is the endpoint name that represents a collection of Ceph monitors
+	EndpointsName string `json:"endpoints"`
+	// Required: RBDImage is the rados image name
+	RBDImage string `json:"image"`
+	// Required: Filesystem type to mount.
+	// Must be a filesystem type supported by the host operating system.
+	// Ex. "ext4", "xfs", "ntfs"
+	// TODO: how do we prevent errors in the filesystem from compromising the machine
+	FSType string `json:"fsType,omitempty"`
+	// Optional: RBDPool is the rados pool name,default is rbd
+	RBDPool string `json:"pool"`
+	// Optional: RadosUser is the rados user name, default is admin
+	RadosUser string `json:"user"`
+	// Optional: Keyring is the path to key ring for RBDUser, default is /etc/ceph/keyring
+	Keyring string `json:"keyring"`
+	// Optional: Secret is the authentication secret for RBDUser, default is empty.
+	Secret string `json:"secret" description:"secret is authentication secret. If provided, it overrides keyring"`
```

Comment:
Interesting approach.  Do we have any volume types that need multiple secrets?

----- Original Message -----
> > +	EndpointsName string `json:"endpoints"`
> > +	// Required: RBDImage is the rados image name
> > +	RBDImage string `json:"image"`
> > +	// Required: Filesystem type to mount.
> > +	// Must be a filesystem type supported by the host operating system.
> > +	// Ex. "ext4", "xfs", "ntfs"
> > +	// TODO: how do we prevent errors in the filesystem from compromising the
> > machine
> > +	FSType string `json:"fsType,omitempty"`
> > +	// Optional: RBDPool is the rados pool name,default is rbd
> > +	RBDPool string `json:"pool"`
> > +	// Optional: RadosUser is the rados user name, default is admin
> > +	RadosUser string `json:"user"`
> > +	// Optional: Keyring is the path to key ring for RBDUser, default is
> > /etc/ceph/keyring
> > +	Keyring string `json:"keyring"`
> > +	// Optional: Secret is the authentication secret for RBDUser, default is
> > empty.
> > +	Secret string `json:"secret" description:"secret is authentication
> > secret. If provided, it overrides keyring"`
> 
> Am I wrong here or is this kind of really easy to secure any volume plugin?
> 
> What if you're both right?  Users provide a name to a secret in their
> namespace.  Admins provide objectRefs to a Secret.
> 
> ```
> type Volume struct {
>     Name
>     VolumeSource
>     SecretName   // the name of a Secret in the same namespace as the volume
> }
> 
> type PersistentVolumeSpec {
>      PersistentVolumeSource
>      SecretRef   // ObjectRef to any Secret in the system.
> }
> 
> 
> // VolumePlugins
> type Volume interface {
>   GetPath()
>   SetSecret(secret)  // kubelet injects the secret
> }
> 
> ```
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/6578/files#r28806909